### PR TITLE
Ta load capability

### PR DESCRIPTION
**Pull request by dsemenets**

### CODE REVIEW

File: tee_supplicant/CMakeLists.txt [hunk 1]
```diff
@@ -8,5 +8,6 @@ target_include_directories(TEE_SUPPLICANT INTERFACE include)
 
 zephyr_library()
 zephyr_library_sources(src/tee_supplicant.c)
+zephyr_library_sources(src/teec_ta_load.c)
 zephyr_include_directories(include)
-zephyr_include_directories(../public)
\ No newline at end of file
+zephyr_include_directories(../public)
```
Comment (oleksiimoisieiev) [inline tee_supplicant/CMakeLists.txt:13]: please remove extra line

File: tee_supplicant/include/teec_ta_load.h [hunk 2]
```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright (c) 2014, STMicroelectronics International N.V.
```
Comment (oleksiimoisieiev) [inline tee_supplicant/include/teec_ta_load.h]: Are you sure it's correct Copyright?

File: tee_supplicant/include/teec_ta_load.h [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+int TEECI_LoadSecureModule(const TEEC_UUID *destination, void *ta,
+			   size_t *ta_size);
+
+struct ta_table {
+	char *uuid;
+	void *ta_start;
+	size_t ta_size;
+};
+
+void TEEC_SetTATable(struct ta_table *table);
```
Comment (oleksiimoisieiev) [inline tee_supplicant/include/teec_ta_load.h:43]: This requires comment as it is in LoadSecureModule

File: tee_supplicant/src/tee_supplicant.c [hunk 4]
```diff
@@ -21,6 +22,25 @@ LOG_MODULE_REGISTER(tee_supplicant);
 static struct k_thread main_thread;
 static K_THREAD_STACK_DEFINE(main_stack, 8192);
 
+struct suppl_shm {
+	int id;
+	void *p;
+	size_t size;
+	bool registered;
+	int fd;
```
Comment (oleksiimoisieiev) [inline tee_supplicant/src/tee_supplicant.c]: why do you need fd here?

File: tee_supplicant/src/tee_supplicant.c [hunk 5]
```diff
@@ -62,10 +88,134 @@ static int send_response(const struct device *dev, struct tee_supp_response *rsp
...
+		 * Buffer isn't found, the caller is querying required size
+		 * of the buffer.
+		 */
+		return 0;
+	}
+
+	sz = MEMREF_SIZE(params + idx);
+	offs = MEMREF_SHM_OFFS(params + idx);
+	if ((sz + offs) < sz)
+		return -1;
```
Comment (oleksiimoisieiev) [inline tee_supplicant/src/tee_supplicant.c]: Please return at least errno on error

File: tee_supplicant/src/tee_supplicant.c [hunk 6]
```diff
@@ -124,4 +274,26 @@ int tee_supp_init(const struct device *dev)
 	return 0;
 }
 
+void tee_supp_mutex_lock(struct k_mutex *mu)
+{
+	int e = k_mutex_lock(mu, K_FOREVER);
+
+	if (e) {
+		LOG_ERR("pthread_mutex_lock: %d", e);
+//		LOG_ERR("terminating...");
```
Comment (oleksiimoisieiev) [inline tee_supplicant/src/tee_supplicant.c]: please no comments

File: tee_supplicant/src/tee_supplicant.c [hunk 7]
```diff
@@ -124,4 +274,26 @@ int tee_supp_init(const struct device *dev)
...
+	}
+}
+
+void tee_supp_mutex_unlock(struct k_mutex *mu)
+{
+	int e = k_mutex_unlock(mu);
+
+	if (e) {
+		LOG_ERR("pthread_mutex_unlock: %d", e);
+//		LOG_ERR("terminating...");
```
Comment (oleksiimoisieiev) [inline tee_supplicant/src/tee_supplicant.c]: please no comments

File: tee_supplicant/src/teec_ta_load.c [hunk 8]
```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright (c) 2014, STMicroelectronics International N.V.
```
Comment (oleksiimoisieiev) [inline tee_supplicant/src/teec_ta_load.c:2]: copyright?

File: tee_supplicant/CMakeLists.txt [hunk 9]
```diff
@@ -8,5 +8,6 @@ target_include_directories(TEE_SUPPLICANT INTERFACE include)
 
 zephyr_library()
 zephyr_library_sources(src/tee_supplicant.c)
+zephyr_library_sources(src/teec_ta_load.c)
```
Comment (oleksiimoisieiev) [inline tee_supplicant/CMakeLists.txt:11]: Commit descrition

File: tee_supplicant/src/teec_ta_load.c [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+#include <teec_ta_load.h>
+
+LOG_MODULE_REGISTER(teec_ta_load);
+#define BINARY_PREFIX ta_load
+#include <teec_trace.h>
+static struct ta_table *ta_table = NULL;
+
+void TEEC_SetTATable(struct ta_table *table)
+{
+	ta_table = table;
```
Comment (oleksiimoisieiev) [inline tee_supplicant/src/teec_ta_load.c:44]: if (table != NULL)?

File: tee_supplicant/src/teec_ta_load.c [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+		     destination->clockSeqAndNode[1],
+		     destination->clockSeqAndNode[2],
+		     destination->clockSeqAndNode[3],
+		     destination->clockSeqAndNode[4],
+		     destination->clockSeqAndNode[5],
+		     destination->clockSeqAndNode[6],
+		     destination->clockSeqAndNode[7]);
+
+	for (t = ta_table; t->uuid != NULL; t++)
+		if (!strcmp(uuid, t->uuid)) {
```
Comment (oleksiimoisieiev) [inline tee_supplicant/src/teec_ta_load.c]: please prefer using strncmp

File: tee_supplicant/src/teec_ta_load.c [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+	ta_table = table;
+}
+
+int TEECI_LoadSecureModule(const TEEC_UUID *destination, void *ta, size_t *ta_size)
+{
+	int res = TA_BINARY_NOT_FOUND;
+	struct ta_table *t;
+	char uuid[PATH_MAX] = { 0 };
+
+	if (!ta || !ta_table)
```
Comment (oleksiimoisieiev) [inline tee_supplicant/src/teec_ta_load.c]: || !ta_size

**Review by oleksiimoisieiev — CHANGES_REQUESTED** [review]

File: tee_supplicant/src/tee_supplicant.c [hunk 13]
```diff
@@ -41,7 +41,7 @@ union tee_supp_msg {
 static int receive_request(const struct device *dev, struct tee_supp_request *ts_req)
 {
 	LOG_ERR("Receiving request...");
```
Comment (oleksiimoisieiev) [inline tee_supplicant/src/tee_supplicant.c]: Please try to avoid this commit

**Review by oleksiimoisieiev — CHANGES_REQUESTED** [review]

File: tee_supplicant/CMakeLists.txt (hunk 1 above)
Comment (dsemenets) [reply tee_supplicant/CMakeLists.txt:13]: This line is not my change

File: tee_supplicant/src/tee_supplicant.c (hunk 4 above)
Comment (dsemenets) [reply tee_supplicant/src/tee_supplicant.c]: removed

File: tee_supplicant/src/tee_supplicant.c (hunk 6 above)
Comment (dsemenets) [reply tee_supplicant/src/tee_supplicant.c]: Done

File: tee_supplicant/src/tee_supplicant.c (hunk 7 above)
Comment (dsemenets) [reply tee_supplicant/src/tee_supplicant.c]: Done

File: tee_supplicant/src/teec_ta_load.c (hunk 8 above)
Comment (dsemenets) [reply tee_supplicant/src/teec_ta_load.c:2]: Fixed

File: tee_supplicant/src/teec_ta_load.c (hunk 10 above)
Comment (dsemenets) [reply tee_supplicant/src/teec_ta_load.c:44]: Why?

File: tee_supplicant/src/teec_ta_load.c (hunk 12 above)
Comment (dsemenets) [reply tee_supplicant/src/teec_ta_load.c]: Fixed

File: tee_supplicant/src/teec_ta_load.c (hunk 11 above)
Comment (dsemenets) [reply tee_supplicant/src/teec_ta_load.c]: Ok

File: tee_supplicant/src/tee_supplicant.c (hunk 5 above)
Comment (dsemenets) [reply tee_supplicant/src/tee_supplicant.c]: Ok

File: tee_supplicant/src/tee_supplicant.c (hunk 13 above)
Comment (dsemenets) [reply tee_supplicant/src/tee_supplicant.c]: Actually it is not my change but Ok

File: tee_supplicant/include/teec_ta_load.h (hunk 3 above)
Comment (dsemenets) [reply tee_supplicant/include/teec_ta_load.h:43]: Done

File: tee_supplicant/include/teec_ta_load.h (hunk 2 above)
Comment (dsemenets) [reply tee_supplicant/include/teec_ta_load.h]: Corrected

File: tee_supplicant/include/teec_ta_load.h (hunk 3 above)
Comment (oleksiimoisieiev) [reply tee_supplicant/include/teec_ta_load.h:43]: +

File: tee_supplicant/src/teec_ta_load.c (hunk 10 above)
Comment (oleksiimoisieiev) [reply tee_supplicant/src/teec_ta_load.c:44]: +

**Review by oleksiimoisieiev — APPROVED** [review]

**Comment by oleksiimoisieiev** [comment]
Add commit message and +1 
`Acked-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>`

**Pull request closed without merging**